### PR TITLE
feat(ship): pr shepherding scaffolding + flag-gated trigger surfaces

### PR DESCRIPTION
**Pull request by chrisleekr**
## Summary

Lands the scaffolding for the PR-shepherding feature (`bot:ship` lifecycle): a probe → verdict → intent → continuation pipeline that drives a PR toward merge-ready under a single durable session per PR. New trigger surfaces (literal `bot:ship`, natural-language mention via Bedrock classifier, and per-command labels) sit behind `SHIP_USE_TRIGGER_SURFACES_V2`; the legacy mention-only handler stays as the default until the flag flips. Adds the `resolve_review_thread` MCP server bound to a single PR for safe thread resolution. Schema changes ship in migration `008_ship_intents.sql` (four tables; partial unique index enforces one in-flight session per PR, covering both `active` and `paused`). Includes the resolution of CodeRabbit's three critical findings and the major+minor batch on top of the initial scaffold (commits `cd9867e`, `32dc584`).

## Diagram

```mermaid
flowchart LR
    classDef oldNode fill:#fef3c7,stroke:#92400e,color:#451a03
    classDef newNode fill:#d1fae5,stroke:#065f46,color:#022c22
    classDef trigger fill:#dbeafe,stroke:#1e3a8a,color:#0c1e3a
    classDef gate fill:#fde68a,stroke:#78350f,color:#451a03

    PRevt["PR / comment / label / review event"]:::trigger
    LegacyMention["Legacy: @mention only<br/>processRequest pipeline"]:::oldNode
    Flag{"SHIP_USE_TRIGGER_SURFACES_V2"}:::gate
    Router["trigger-router<br/>literal / NL / label"]:::newNode
    Eligibility["eligibility check"]:::newNode
    Intent["createIntent<br/>partial unique idx"]:::newNode
    Probe["probe<br/>GraphQL MergeReadiness"]:::newNode
    Verdict["computeVerdict<br/>NonReadinessReason"]:::newNode
    Cont["continuation row<br/>+ ship:tickle ZSET"]:::newNode
    Reactor["webhook reactor<br/>fanOut early-wake"]:::newNode
    Term["terminal: ready_awaiting_human_merge<br/>or human_took_over / pr_closed"]:::newNode

    PRevt --> Flag
    Flag -->|off| LegacyMention
    Flag -->|on| Router
    Router --> Eligibility --> Intent --> Probe --> Verdict
    Verdict -->|ready| Term
    Verdict -->|not ready| Cont
    Cont --> Reactor
    Reactor --> Probe
```

### Spec & contracts (P0)
- `specs/20260427-201332-pr-shepherding-merge-ready/` — full Spec Kit set: `spec.md`, `plan.md`, `research.md`, `data-model.md`, `tasks.md`, `quickstart.md`, plus four contract docs (`probe-graphql-query.md`, `bot-commands.md`, `webhook-event-subscriptions.md`, `mcp-resolve-thread-server.md`, `resolve-thread-mutation.md`).
- `docs/SHIP.md`, `docs/ARCHITECTURE.md`, `docs/CONFIGURATION.md`, `docs/OBSERVABILITY.md`, `docs/SETUP.md`, `docs/EXTENDING.md`, `docs/BOT-WORKFLOWS.md`, `mkdocs.yml` updated to reflect the new module surface.

### Database (P2)
- `src/db/migrations/008_ship_intents.sql` adds `ship_intents`, `ship_iterations`, `ship_continuations`, `ship_fix_attempts`. Partial unique index `WHERE status IN ('active', 'paused')` enforces FR-007a.
- `src/db/queries/ship.ts` typed Bun.sql helpers; `scripts/db-migrate.ts` migration runner.

### Ship workflow (P1–P7)
- Pure verdict + GraphQL probe: `src/workflows/ship/{verdict,probe,review-barrier}.ts` (P1).
- Intent lifecycle + state machine: `src/workflows/ship/intent.ts` (P2) with `pauseIntent`/`resumeIntent`/`forceAbortIntent`/`transitionToTerminal`.
- Continuation + cron tickle: `src/workflows/ship/{continuation,tickle-scheduler}.ts` (P3) with reentrancy guard and Postgres reconciliation on boot.
- Webhook reactor: `src/workflows/ship/webhook-reactor.ts` + new event handlers (`check-run.ts`, `check-suite.ts`) + reactor bridge (P4); per-intent error isolation.
- Fix-attempts + signature + deadline: `src/workflows/ship/{fix-attempts,signature,deadline}.ts` (P5) with two-tier signature derivation.
- Abort + lifecycle commands: `src/workflows/ship/{abort,lifecycle-commands}.ts` (P6) with cooperative-then-forced cancellation and foreign-push fail-closed resume.
- Resolve-review-thread MCP server: `src/mcp/servers/resolve-review-thread.ts` + `registry.ts` registration (P7), bound to one `(owner, repo, pr)` per instance.

### Trigger surfaces (`SHIP_USE_TRIGGER_SURFACES_V2`)
- `src/workflows/ship/{trigger-router,literal-command,nl-classifier,label-trigger,command-dispatch,session-runner,reactor-bridge}.ts`.
- NL classifier reuses the existing Bedrock single-turn adaptor (`src/ai/llm-client.ts`); FR-025a mention-prefix gate runs **before** any LLM call.
- `BOT_LABEL_PATTERN` permits documented label shapes (`bot:abort-ship`, `bot:fix-thread`, `bot:ship/deadline=2h`, …).

### Config + log fields
- `src/config.ts` adds the ship env block (`MAX_WALL_CLOCK_PER_SHIP_RUN`, `CRON_TICKLE_INTERVAL_MS`, `MERGEABLE_NULL_BACKOFF_MS_LIST`, `REVIEW_BARRIER_SAFETY_MARGIN_MS`, `FIX_ATTEMPTS_PER_SIGNATURE_CAP`, `SHIP_USE_PROBE_VERDICT`, `SHIP_USE_CONTINUATION_LOOP`, `SHIP_USE_TRIGGER_SURFACES_V2`, `BOT_APP_LOGIN`).
- `src/workflows/ship/log-fields.ts` Zod-validated structured-log schema (FR-016).

### CI safety
- `scripts/check-no-destructive-actions.ts` greps the ship subtree for force-push / reset-hard / merge-API patterns; fails closed on filesystem errors.

### CodeRabbit follow-ups in this PR
- 3 critical (`cd9867e`): cross-PR resolve-thread preflight, eligibility owner-allowlist single-principal check, tickle-scheduler dispatch retry.
- Major+minor batch (`32dc584`): ANSI regex fix, label pattern broaden, NL prefix gate, reentrancy guard, double-`@` reply, hardcoded trigger phrase removed, resume fail-closed, reactor isolation, structured `unique_violation` detection (Bun.sql `errno`/`constraint`), flake-tracker pino logging, `databaseId` for projected reruns, review-barrier `__typename` filter, pause/resume guarded reply, plus the C-bucket doc/spec sweep.

## Related Issues

- Closes #N/A — feature scaffold (no tracking issue)

## Test plan

- [x] Tested locally — `bun run typecheck` clean, `bun run lint` zero errors
- [x] Added/updated tests — 98 ship-suite tests pass (probe / verdict / intent / lifecycle / signature / fix-attempts / cancellation / eligibility / review-barrier / session-runner / reactor / migration `008.test.ts`)
- [x] All existing tests pass (pre-existing skipped opt-in suite `repo-knowledge.test.ts` and unrelated K8s/runs-store/intent-classifier failures predate this branch — verified via `git stash` baseline)
- [ ] Smoke-test under `SHIP_USE_TRIGGER_SURFACES_V2=true` in dev installation before flag flip
- [ ] Verify migration `008_ship_intents.sql` applies cleanly on staging Postgres

## Release Notes

* **New Features**
  * Added PR shepherding workflow (`bot:ship`) with automated merge-readiness probing and iterative state management
  * Enabled three equivalent command invocation surfaces: literal `bot:ship` comments, natural language mentions, and GitHub labels
  * Added lifecycle control commands: `bot:stop`, `bot:resume`, `bot:abort-ship`
  * Added resolve-review-thread capability for managing PR review feedback
  * Enhanced webhook subscriptions for PR synchronization, CI completion events, and review activity

* **Documentation**
  * Comprehensive operator guides for PR shepherding workflows, configuration, setup, and architecture
  * Added observability and logging contract specifications

* **Tests**
  * Extensive integration and unit test coverage for new features and workflows

### CODE REVIEW

File: src/mcp/servers/resolve-review-thread.ts [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+              text: JSON.stringify({
+                code: "graphql_error",
+                message: `thread belongs to PR #${String(thread.pullRequest.number)} but this server is bound to PR #${String(BOUND_PR_NUMBER)}`,
+                thread_id,
+              }),
+            },
+          ],
+          isError: true,
+        };
+      }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/mcp/servers/resolve-review-thread.ts]: Fixed in cd9867e: added a read-only `GetReviewThread` preflight query that fetches the thread's pull-request number, and the `resolveReviewThread` mutation now only fires when that number matches `BOUND_PR_NUMBER`. Cross-PR threads are rejected before any side effect.

File: src/workflows/ship/eligibility.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      (o) => o.toLowerCase() === input.triggeringUserLogin.toLowerCase(),
+    );
+    const ownerOk = allowedOwners.some((o) => o.toLowerCase() === input.owner.toLowerCase());
+    if (!triggerOk && !ownerOk) {
+      return {
+        eligible: false,
+        reason: "unauthorized",
+        message: `\`${input.triggeringUserLogin}\` is not in the configured ALLOWED_OWNERS list.`,
+      };
+    }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/eligibility.ts:78]: Fixed in cd9867e: removed the `ownerOk` variable and the fallback. The gate now checks only the triggering principal against `ALLOWED_OWNERS`, matching the FR-015/FR-028 wording and the existing `unauthorized` error message which already only references `triggeringUserLogin`.

File: src/workflows/ship/tickle-scheduler.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    ])) as string[] | null;
+    if (due === null || due.length === 0) return;
+    for (const intent_id of due) {
+      await deps.valkey.send("ZREM", [TICKLE_KEY, intent_id]);
+      try {
+        await deps.onDue(intent_id);
+      } catch (err) {
+        logger.error(
+          { event: "ship.tickle.dispatch_failed", intent_id, err: String(err) },
+          "ship tickle dispatch failed",
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/tickle-scheduler.ts]: Fixed in cd9867e: on `onDue` failure the intent is now re-armed via `ZADD` with a one-interval delay so the session stays live without tight-looping a known-bad dispatch. Reconciliation on restart still serves as the backstop, but no longer the only recovery path.

**Comment by chrisleekr** [comment]
@coderabbitai — addressed your major+minor batch in 32dc584:

**Fixed (A bucket — correctness/security):** nl-classifier prefix-only mention gate, signature.ts ANSI escape regex (added `\x1b`), pull-request `BOT_LABEL_PATTERN` broadened for `bot:abort-ship` / `/deadline=...`, tickle-scheduler reentrancy guard, double-`@` in already-in-progress reply, hardcoded trigger phrase in tracking-comment, lifecycle-commands resume fail-closed, webhook-reactor per-intent error isolation, intent.ts structured unique-violation detection (Bun.sql exposes SQLSTATE on `errno`, constraint name on `constraint`), flake-tracker pino logging on rerun failure, GraphQL `databaseId` so projected probe history can populate `check_run_id`, `check-no-destructive-actions` walk() fails closed.

**Fixed (B bucket — config-driven constants):** new `botAppLogin` config field + env override `BOT_APP_LOGIN`; review.ts log now scoped to the installation branch.

**Fixed (D — targeted):** new `review_barrier_deferred` non-readiness reason (no longer misencodes the latency gate as `human_took_over`); command-dispatch try/catch around classifier; eligibility test fixture uses `state: "MERGED"`; pause/resume replies now reflect the guarded UPDATE result; review-barrier carries `author.__typename` and excludes any non-`User` actor; tracking-comment marker is patched to the canonical `<!-- ship-intent:{id} -->` form post-create.

**Fixed (C — docs/spec):** spec.md inlines the architecture decisions (removes the private Dropbox path); probe-graphql-query.md, mcp-resolve-thread-server.md, webhook-event-subscriptions.md, data-model.md, tasks.md, research.md, quickstart.md, SETUP.md, EXTENDING.md, log-fields.ts comment, db/queries/ship.ts comment — all aligned.

**Skipped with rationale (in commit message):**
- `probe.ts` retry → shared util: `mergeableNullBackoffMsList` is a fixed schedule, not exponential — the shared util's validated parameters don't fit.
- `issue-comment.ts` dual-dispatch dedup: `SHIP_USE_TRIGGER_SURFACES_V2` is being removed shortly; the dedup guard would be throwaway.
- `verdict.test.ts` missing-PR test relocation: the underlying verdict behaviour for null PR is consistent with terminal-state handling; moving the test alone doesn't address the semantic concern and changing the verdict reason cascades outside this PR's scope.

`bun run typecheck` clean, `bun run lint` zero errors, all 98 ship tests pass.

**Comment by chrisleekr** [comment]
@coderabbitai — round 2 addressed in cbf4259.

**Critical (1):**
- `mcp/resolve-review-thread.ts` preflight now validates `(owner, repo, number)` not just `number` — added `pullRequest.repository.{owner.login,name}` to `GET_THREAD_QUERY` and rejects any drift from the bound tuple.

**Major code (7):**
- `lifecycle-commands.ts` foreign-push detection: switched from `pr.head.user` (head-repo owner per the GitHub REST contract) to `octokit.rest.repos.getCommit(ref=head.sha)` and compared `commit.author.login` (with committer fallback) against `config.botAppLogin`. Test mock updated.
- `nl-classifier.ts`: added a token-boundary check after the trigger phrase so `@chrisleekr-bot-foo` no longer slips through.
- `signature.ts`: re-anchored the Tier-1 ESLint regex on EOL and added scoped-rule support (`@scope/name`).
- `mcp/resolve-review-thread.ts`: both GraphQL calls wrapped in `retryWithBackoff`; both error paths now include `pr_number`; `runServer().catch` exits non-zero on transport failure.
- `check-no-destructive-actions.ts`: forbidden patterns broadened (plain `git reset --hard`, `git push --delete`, case-insensitive `branch -D`).

**Major docs (2):**
- `contracts/probe-graphql-query.md` reviewer rule now matches `review-barrier.ts` (`__typename === 'User'`, no App allowlist).
- `data-model.md` `non_readiness_reason` CHECK + `NON_READINESS_REASONS` list now include `review_barrier_deferred`.

**Minor (5):** SETUP.md handler-status note, plan.md "markered" typo, pull-request.ts docstring, command-dispatch `err` preservation, eligibility tests for `unauthorized` + `forbidden_target_branch` (mutate config singleton since `loadConfig()` runs once at module load).

**Trivials (4):** probe.ts `as never` → typed cast via imported `BarrierProbeShape`; probe.ts audit-row catch now logs at `warn`; reactor-bridge logs `debug` when `DATABASE_URL` is absent; verdict.ts `import { z }` moved to top-of-file.

**Pushing back (2) — documented in commit message:**
- `check-no-destructive-actions.ts:80` multi-line scanning: line-by-line is intentional. The guard flags obvious shell forms; multi-line normalisation expands the false-positive surface (block comments, JSDoc, prompt strings) without meaningfully improving safety. Tool-call recorder coverage exists in T046a for the agent-execution path.
- `docs/EXTENDING.md:230` handler path: the suggested `src/workflows/ship/handlers/resolve.ts` does not exist — actual handler is at `src/workflows/handlers/resolve.ts` (`grep enableResolveReviewThread src/`). Current docs are correct.

`bun run typecheck` clean, `bun run lint` zero errors, `bun test test/workflows/ship/ test/mcp/` → 100 pass, 0 fail.

File: specs/20260427-201332-pr-shepherding-merge-ready/contracts/probe-graphql-query.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+          comments(first: 1) {
+            nodes {
+              author {
+                login
+              }
+              createdAt
+            }
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260427-201332-pr-shepherding-merge-ready/contracts/probe-graphql-query.md:44]: Fixed in 1b1924e: dropped the unused `reviewThreads.nodes[].comments` selection, added `CheckRun.databaseId`, and added `reviewThreads.pageInfo { hasNextPage endCursor }` so the contract now matches the shipped probe query.

File: specs/20260427-201332-pr-shepherding-merge-ready/data-model.md [hunk 5]
```diff
@@ -0,0 +1,371 @@
...
+  | {
+      ready: false;
+      reason: NonReadinessReason;
+      detail: string;
+      checked_at: string;
+      head_sha: string;
+    };
+```
+
+All three enumerations are SQL `CHECK` constraints AND TS literal-union types AND Zod enums (validated at every boundary).
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply specs/20260427-201332-pr-shepherding-merge-ready/data-model.md:289]: Fixed in 1b1924e: added `"review_barrier_deferred"` to the `NonReadinessReason` TS union in `src/db/queries/ship.ts`, so the SQL `CHECK` constraint, TS literal-union, and Zod enum are now genuinely in lockstep.

File: src/mcp/servers/resolve-review-thread.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  REPO_OWNER === "" ||
+  REPO_NAME === undefined ||
+  REPO_NAME === "" ||
+  PR_NUMBER === undefined ||
+  PR_NUMBER === "" ||
+  GITHUB_TOKEN === undefined ||
+  GITHUB_TOKEN === ""
+) {
+  console.error("Error: REPO_OWNER, REPO_NAME, PR_NUMBER, and GITHUB_TOKEN are required");
+  process.exit(1);
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/mcp/servers/resolve-review-thread.ts]: Fixed in 1b1924e: replaced all three `console.error` call sites with structured pino logging. The MCP server uses a locally-instantiated pino logger bound to `process.stderr` rather than importing `src/logger.ts` — the project root logger writes to stdout, which would corrupt the JSON-RPC stream this stdio-transport sidecar uses for tool calls.

File: src/mcp/servers/resolve-review-thread.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+  const status = (err as { status?: number }).status;
+  if (status === 404) return "thread_not_found";
+  if (status === 403) return "permission_denied";
+  if (status === 429) return "rate_limited";
+
+  const message = err instanceof Error ? err.message.toLowerCase() : String(err).toLowerCase();
+  if (message.includes("not found")) return "thread_not_found";
+  if (message.includes("forbidden") || message.includes("permission")) return "permission_denied";
+  if (message.includes("rate limit") || message.includes("secondary rate")) return "rate_limited";
+  if (
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/mcp/servers/resolve-review-thread.ts:143]: Fixed in 1b1924e: `classifyError` now normalises the error message before the status-code branches, and inspects for `rate limit`/`secondary rate` substrings ahead of the 403 check. Secondary rate-limit 403s now map to `rate_limited`; only true permission-denied 403s fall through to `permission_denied`.

File: src/webhook/events/pull-request.ts [hunk 8]
```diff
@@ -26,6 +36,33 @@ export function handlePullRequest(
...
+      type: "pull_request.synchronize",
+      installation_id: payload.installation.id,
+      owner: payload.repository.owner.login,
+      repo: payload.repository.name,
+      pr_number: payload.pull_request.number,
+      head_sha: payload.pull_request.head.sha,
+      head_author_login: payload.sender.login,
+    });
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/webhook/events/pull-request.ts:43]: Fixed in 1b1924e: extracted a new `handlePullRequestSynchronize` helper that calls `octokit.rest.repos.getCommit({ ref: head.sha })` and resolves `head_author_login` from `commit.author?.login ?? commit.committer?.login` (with `payload.sender.login` retained as a last-resort fallback if the lookup fails). This now matches the lifecycle-commands.ts pattern from round 2 — cherry-picks, rebases, and bot-pushed-on-behalf-of flows are now classified by actual commit author rather than webhook actor.

File: src/workflows/ship/probe.ts [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        reviewDecision
+        baseRefName
+        baseRefOid
+        headRefName
+        headRefOid
+        author { login }
+        reviewThreads(first: 100) {
+          totalCount
+          nodes { id isResolved isOutdated }
+        }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/probe.ts:50]: Fixed in 1b1924e: added a dedicated `REVIEW_THREADS_PAGE_QUERY` and an `ensureFullReviewThreads` helper that paginates with `pageInfo.hasNextPage`/`endCursor`. Pagination short-circuits as soon as it sees an unresolved + non-outdated thread (the verdict only needs to know whether at least one exists), and any GraphQL failure during pagination propagates so the probe fails closed rather than silently treating a truncated response as 'ready'.

File: src/workflows/ship/probe.ts [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+  const backoff = input.mergeableBackoffMs ?? config.mergeableNullBackoffMsList;
+  const sleep = input.sleep ?? defaultSleep;
+
+  let lastResponse: ProbeResponseShape | null = null;
+  for (let attempt = 0; attempt <= backoff.length; attempt += 1) {
+    const response = await input.octokit.graphql<ProbeResponseShape>(PROBE_QUERY, {
+      owner: input.owner,
+      repo: input.repo,
+      number: input.pr_number,
+    });
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/probe.ts]: Fixed in 1b1924e: every individual `octokit.graphql` call inside `runProbe` (and the new `paginateReviewThreads`) is now wrapped in `retryWithBackoff` from `src/utils/retry.ts` (3 attempts, exponential backoff). The outer mergeable=null backoff loop is left intact because it has different semantics (waiting for GitHub to finish computing mergeable, not retrying a transient failure) — that distinction is now called out in a code comment too.

File: src/workflows/ship/signature.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    .toLowerCase();
+}
+
+function tryExtractTier1(text: string): string | null {
+  for (const re of TIER1_PATTERNS) {
+    const match = re.exec(text);
+    if (match !== null) {
+      const groups = match.slice(1).filter((g): g is string => typeof g === "string");
+      if (groups.length > 0) return groups.join("|");
+    }
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/signature.ts]: Fixed in 1b1924e along the lines of your suggested diff: `TIER1_PATTERNS` is now `readonly Tier1Pattern[]` where each entry is `{ re, stableGroup }`. Each regex captures only the stable identifier (rule id, TS error code, error class); volatile message text is matched as `.+?` but not captured. `tryExtractTier1` returns just `match[stableGroup]`, so two failures of the same rule on different lines/files now share a signature and the fix-attempts cap counts retries correctly.

File: src/workflows/ship/verdict.ts [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+        ctx.status === "QUEUED" ||
+        ctx.status === "IN_PROGRESS" ||
+        ctx.status === "PENDING"
+      ) {
+        pendingRequired.push(ctx.name ?? "<unknown>");
+      }
+    } else {
+      const s = ctx.state;
+      if (s === "FAILURE" || s === "ERROR") failingRequired.push(ctx.context ?? "<unknown>");
+      else if (s === "PENDING") pendingRequired.push(ctx.context ?? "<unknown>");
```
[the remark this answers is not in the document]
Comment (chrisleekr) [reply src/workflows/ship/verdict.ts]: Fixed in 1b1924e: the StatusContext branch in `computeVerdict` now treats `s === "PENDING" || s === "EXPECTED"` as pending. A short comment in the code documents the GitHub StatusState enum (`EXPECTED | PENDING | FAILURE | ERROR | SUCCESS`) and why `EXPECTED` legacy required checks must block ready.
